pkg/leech/pipeline.go

```diff
@@ -0,0 +1,269 @@
+		if errors.Is(err, errLogsExpired) {
+			logger.Infof("logs for workflow not available", "DeliveryID", event.DeliveryID)
+			result.Status = "NOT_FOUND"
+		} else {
+			// Other failures can be retried
+			// @TODO(bradegler): These are currently filtered out of the query, need to figure
+			// out a way to mark how many attempts have been made for a particular delivery id
+			// preferably without causing an update to the row in BigQuery. The simplest approach
+			// would be to just not write a FAILURE status and let the query try again. The drawback
+			// is that something could be stuck forever in that situation and we wouldn't want to
+			// keep processing it. I think a secondary FAILURE table might work that we can join
+			// into the main query as WHERE count(failures where delivery_id = x) < 10 or something.
+			// This adds complexity to the write operation though so it requires some thought.
+			// For now just flag rows as FAILUREs and we can delete them from the table to trigger
+			// reprocessing.
+			logger.Errorf("failed to retrieve logs for workflow: %w", err, "DeliveryID", event.DeliveryID)
+			result.Status = "FAILURE"
+		}
+	}
+	return result
+}
+
+// handleMessage is the main event processor. It generates a GitHub token, reads the workflow
+// log files if they exist and persists them to Cloud Storage.
+func (f *ingestLogsFn) handleMessage(ctx context.Context, repoName, ghLogsURL, gcsPath string) error {
+	tokenRequest := githubapp.TokenRequest{
+		Repositories: []string{repoName},
+		Permissions: map[string]string{
+			"actions": "read",
+		},
+	}
+	// @TODO(bradegler): This could use some caching. Requests to the same repo
+	// can reuse a token without requesting a new one until it expires. Might be
+	// better to implement that in pkg so that GHTM can take advantage of it as well.
+	ghAppJWT, err := f.pipe.ghApp.AccessToken(ctx, &tokenRequest)
+	if err != nil {
+		return fmt.Errorf("error creating GitHub access token: %w", err)
+	}
+	// The token response is a json doc with a lot of information about the
+	// token. All that is needed is the token itself.
+	var tokenResp map[string]any
+	if err := json.Unmarshal([]byte(ghAppJWT), &tokenResp); err != nil {
+		return fmt.Errorf("malformed GitHub token response: %w", err)
+	}
+	token, ok := tokenResp["token"].(string)
+	if !ok {
+		return fmt.Errorf("malformed GitHub token: wanted string got %T", token)
+	}
+
+	// Create a request to the workflow logs endpoint. This will follow redirects
+	// by default which is important since the endpoint returns a 302 w/ a short lived
+	// url that expires.
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, ghLogsURL, nil)
+	if err != nil {
+		return fmt.Errorf("error creating http request: %w", err)
+	}
+	req.Header.Set("Accept", "application/vnd.github+json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+
+	client := http.Client{Timeout: 5 * time.Minute}
```

Comment:
Can we create this client once during init and store it on ingestLogsFn?

Reply:
yes